### PR TITLE
feat: face mask for kirakira effect #86epjek8p

### PR DESCRIPTION
**Pull request by ChiaoteNi**
# Description

- Currently, the kirakira effects have a high chance of displaying near the eyes and teeth, which are mostly lighter than its nearby area, and it causes the result to look scary for those cases.
- Therefore, this PR adds face masks to the kirakira effect to prevent the center of sparkles located on people's faces.

# Screenshots

<table>
<tr>
<th>Before</th>
<th>After</th>
</tr>
<tr>
<td>

https://github.com/cardinalblue/pic-collage-ios/assets/40178645/29eb7ee3-38ad-4c52-bcbc-9f0c3978c28b

</td>
<td>

https://github.com/cardinalblue/pic-collage-ios/assets/40178645/ed0c7ccf-2382-4622-ad80-1ee440b8bda0

</td>
</tr>
</table>

### CODE REVIEW

**Review by Vivienne0206 — APPROVED** [review]

File: framework/Source/Sparkles.swift [hunk 1]
```diff
@@ -28,6 +28,18 @@ public class Sparkles: OperationGroup {
         didSet { perlinNoiseEffect.time = time }
     }
     // LightExtractor
+    public var faceMaskImage: UIImage? {
+        didSet {
+            guard faceMaskImage != oldValue else { return }
+            let maskImage: UIImage = faceMaskImage ?? .emptyMaskImage
```
Comment (ChiaoteNi) [inline framework/Source/Sparkles.swift]: Hi @yyjim, as per our previous discussion in another PR: https://github.com/cardinalblue/GPUImage3/pull/23#discussion_r1645232634

I've updated the implementation to handle the nil faceMaskImage case.
Please help to review the PR again when you're available, thanks!

File: framework/Source/Sparkles.swift [hunk 2]
```diff
@@ -197,4 +212,31 @@ extension Sparkles {
...
+            await maskInput.processImage()
+            semaphore.signal()
+        }
+        semaphore.wait()
+    }
+}
+
+private extension UIImage {
+    static func makeEmptyMaskImage() -> UIImage {
+        let size = CGSize(width: 1, height: 1)
```
Comment (ChiaoteNi) [inline framework/Source/Sparkles.swift:213]: Hi @yyjim here's the result when I pass nil as the maskImage:
The size CGSize(width: 1, height: 1) works as expected!

<table>
<tr>
<th>With maskImage</th>
<th>maskImage = nil </br>(makeEmptyMaskImage())</th>
</tr>
<tr>
<td>

https://github.com/cardinalblue/GPUImage3/assets/40178645/aaef08a9-4a1e-4fa2-ae85-15cd63e13fc2

</td>
<td>

https://github.com/cardinalblue/GPUImage3/assets/40178645/794e8ecf-8d7d-40e3-ac89-44dcc9268e26

</td>
</tr>
</table>

**Comment by ChiaoteNi** [comment]
Hi @yyjim, I decided to change the implementation with the way I used for CBAddBlend and CBScreenBlend, and I'll make the PR ready for review again once I push the commit; thanks!

File: framework/Source/Sparkles.swift [hunk 3]
```diff
@@ -197,4 +212,31 @@ extension Sparkles {
...
+        let color = UIColor.clear
```
Comment (yyjim) [inline framework/Source/Sparkles.swift:214]: question: So, an empty mask is a clear color? The mask should be black and white, right?

The old implementation was black.

**Review by yyjim — APPROVED** [review]
please check the comment about empty mask

File: framework/Source/Sparkles.swift (hunk 3 above)
Comment (ChiaoteNi) [reply framework/Source/Sparkles.swift:214]: Yes the empty mask is a clear color; the old implementation used the r channel as the mask, but we decided to change to use the alpha channel at last.
```Metal
outputColor *= (1.0 - voidMaskColor.a);
```

**Comment by ChiaoteNi** [comment]
Hi @yyjim I'm going to merge the PR first, please feel free to leave comments, and I'll create another PR for them if needed
Thanks!

File: framework/Source/Sparkles.swift (hunk 3 above)
Comment (ChiaoteNi) [reply framework/Source/Sparkles.swift:214]: @yyjim  Wait... you meant the previous implementation in this PR
Sorry that was just my mistake 🙏🏻
I remember it wrong with the first version of the face mask

**Pull request merged**